slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/AttestationEpochManager.java

```diff
@@ -0,0 +1,119 @@
+  private final Optional<UInt64> highestTargetEpoch;
+  private final Handle h;
+  private final Validator validator;
+
+  final OptionalMinValueTracker minSourceTracker = new OptionalMinValueTracker();
+  final OptionalMinValueTracker minTargetTracker = new OptionalMinValueTracker();
+
+  private AttestationEpochManager(
+      final LowWatermarkDao lowWatermarkDao,
+      final Optional<UInt64> highestSourceEpoch,
+      final Optional<UInt64> highestTargetEpoch,
+      final Validator validator,
+      final Handle h) {
+    this.lowWatermarkDao = lowWatermarkDao;
+    this.highestSourceEpoch = highestSourceEpoch;
+    this.highestTargetEpoch = highestTargetEpoch;
+    this.h = h;
+    this.validator = validator;
+  }
+
+  public static AttestationEpochManager create(
+      final SignedAttestationsDao signedAttestationsDao,
+      final LowWatermarkDao lowWatermarkDao,
+      final Validator validator,
+      final Handle h) {
+    final Optional<UInt64> maxTargetEpoch =
+        signedAttestationsDao.maximumTargetEpoch(h, validator.getId());
+    final Optional<UInt64> maxSourceEpoch =
+        signedAttestationsDao.maximumSourceEpoch(h, validator.getId());
+
+    return new AttestationEpochManager(
+        lowWatermarkDao, maxSourceEpoch, maxTargetEpoch, validator, h);
+  }
+
+  public void trackEpochs(final UInt64 importedSourceEpoch, final UInt64 importedTargetEpoch) {
+    minSourceTracker.trackValue(importedSourceEpoch);
+    minTargetTracker.trackValue(importedTargetEpoch);
+  }
+
+  public void persist() {
+    final Optional<SigningWatermark> existingWatermark =
+        lowWatermarkDao.findLowWatermarkForValidator(h, validator.getId());
+
+    final Optional<UInt64> newSourceWatermark =
+        findBestSourceEpochWatermark(
+            existingWatermark.flatMap(
+                watermark -> Optional.ofNullable(watermark.getSourceEpoch())));
+    final Optional<UInt64> newTargetWatermark =
+        findBestTargetEpochWatermark(
+            existingWatermark.flatMap(
+                watermark -> Optional.ofNullable(watermark.getTargetEpoch())));
+
+    if (newSourceWatermark.isPresent() && newTargetWatermark.isPresent()) {
+      LOG.info(
+          "Updating validator {} source epoch to {}",
+          validator.getPublicKey(),
+          newSourceWatermark.get());
+      LOG.info(
+          "Updating validator {} target epoch to {}",
+          validator.getPublicKey(),
+          newTargetWatermark.get());
+      lowWatermarkDao.updateEpochWatermarksFor(
+          h, validator.getId(), newSourceWatermark.get(), newTargetWatermark.get());
+    } else if (newSourceWatermark.isPresent() != newTargetWatermark.isPresent()) {
+      throw new RuntimeException(
+          "Inconsistent data - no existing attestation watermark, "
+              + "and import only sets one epoch");
+    }
+  }
+
+  private Optional<UInt64> findBestSourceEpochWatermark(final Optional<UInt64> currentWatermark) {
```

Comment:
This and the findBestTargetEpochWatermark are almost the same is there anyway to remove some of the duplication?